Implements file reading for peer tags to synchronize cross repo usage

zarirhamza opened the pull request:
### What does this PR do?

Make the `defaultPeerTags` list read from a file so that file can be synchronized with other repos that use peer tags

Part of a larger set of changes to come where common tag names are sourced from a single source of truth across DD repos 

The planned source of truth will generate static files which can be read by the agent code and synchronized across all DD repos involving peer tags

### How to test

Ensure that service map [behavior](https://docs.datadoghq.com/tracing/services/services_map/) is unchanged namely the features related to `peer.*` tags. Specifically, the connections between services in the service map remain unchanged


Run `TestPreparePeerTags` in `concentrator_test.go`

zarirhamza committed:
Changes defaultPeerTagList to be read from a file

Part of larger changes to come where common tag names are sourced from a single source of truth across DD repos
The planned source of truth will generate static files which can be read by the agent code and synchronized across all DD repos involving peer tags

zarirhamza commented:
Probably going to shift away from the csv style but for now a place holder until @majorgreys decides on the final expected format to read values from

zarirhamza committed:
linter fixes

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
linter build fixes

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
Adds parsing for ini file

Goal is to have values stored in ini file which is now parsed as expected

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
Adds conditions to ensure tag list matches current behavior

Due to mismatches between remapping on backend and previous list, several tags are inserted and removed to ensure the final list is the same as before
This may change in the future in order to ensure all behavior is consistent
Tests are also added to ensure similar behavior

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza commented:
Final format style is determined to be INI. PR is finished and ready for review but note there are a few mismatches between the backend and agent aggregation. To avoid any breaking changes, some tags are manually added and removed but that may change in the future once discussed with the stats team

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

pkalmakis commented on pkg/trace/go.mod:
```diff
@@ -45,7 +45,8 @@ require (
 	google.golang.org/grpc v1.63.2
 	google.golang.org/protobuf v1.34.0
-	k8s.io/apimachinery v0.25.5
+	gopkg.in/ini.v1 v1.67.0
+  k8s.io/apimachinery v0.25.5
```
nit: looks like you have a whitespace issue on line 49

pkalmakis commented on pkg/trace/stats/concentrator.go:
```diff
@@ -55,37 +58,36 @@ type Concentrator struct {
+	cfg, _ := ini.Load(peerTagFile)
+	keys := cfg.Section("dd.apm.peer.tags").Keys()
+
+	var tags []string = []string{"_dd.base_service", "cassandra.cluster",
+		"hazelcast.instance", "rulename", "statemachinename"}
+
+	if len(keys) > 0 {
+		for _, key := range keys {
+			value := strings.Split(key.Value(), ",")
+			for _, tag := range value {
+				if tag != "grpc.host" && tag != "http.host" &&
+					tag != "cassandra.keyspace" && tag != "rpc.system" &&
+					tag != "messaging.system" && tag != "peer.cluster.name" &&
+					tag != "hostname" && tag != "out.host" &&
+					tag != "messaging.destination" && tag != "messaging.destination.name" &&
+					tag != "messaging.rabbitmq.exchange" {
+					tags = append(tags, tag)
+				}
+			}
+		}
+	}
```
This doesn't make sense to me. Why are we defining the tags in a file, but then hardcoding all of these defaults and exceptions??

djova commented on pkg/trace/stats/peer_tags.ini:
```diff
@@ -0,0 +1,22 @@
+peer.couchbase.seed.nodes = "db.couchbase.seed.nodes"
+peer.db.name = "db.name,mongodb.db,db.instance,cassandra.keyspace"
+peer.db.system = "db.system"
+peer.hostname = "peer.hostname,hostname,net.peer.name,db.hostname,network.destination.name,grpc.host,http.host,server.address,out.host"
```
Let's remove `out.host` we actually don't want this one aggregated by default in the agent.

zarirhamza committed:
Updates generated file to prevent mismatch between aggregation and backend remapping

After consulting stats team, new INI file is used to generate tags to aggregate on

zarirhamza committed:
Update go.mod

spacing issue on go.mod

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
Updates list to include new otel peer source tags

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

pkalmakis commented on pkg/trace/stats/concentrator.go:
```diff
@@ -55,37 +58,26 @@ type Concentrator struct {
+//go:embed peer_tags.ini
+var peerTagFile []byte
+
+var defaultPeerTags = func() []string {
```
nit: in my experience, it's more idiomatic `go` to set the value of a global var in an `init` function. I don't know that there is a difference in using an anonymous function like this, but I've never seen it done this way before.

pkalmakis reviewed (approved):
A couple of nitpicks, but otherwise LGTM

zarirhamza committed:
removes unnecessary length check

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

knusbaum commented on pkg/trace/stats/concentrator.go:
```diff
@@ -55,37 +58,24 @@ type Concentrator struct {
+var peerTagFile []byte
+
+var defaultPeerTags = func() []string {
+	cfg, _ := ini.Load(peerTagFile)
```
We need to check for an error here and provide some sort of more gentle failure. Otherwise if there's any error loading or parsing the file we'll just panic with a `nil` pointer error.

zarirhamza committed:
Added error handling for parsing INI file

Synced go.mod files

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

knusbaum reviewed (approved):
My comments have been addressed.

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
fix go.mod files

songy23 reviewed (commented):
Could you do a quick sanity check on whether OTel collector can still build with this change? Basically
1. get a pseudo version of pkg/trace & comp/otelcol/otlp/components/statsprocessor
2. use it in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/go.mod#L25
3. run `make otelcontribcol`

zarirhamza commented:
> Could you do a quick sanity check on whether OTel collector can still build with this change? Basically
> 
> 1. get a pseudo version of pkg/trace & comp/otelcol/otlp/components/statsprocessor
> 2. use it in https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/datadogexporter/go.mod#L25
> 3. run `make otelcontribcol`

Ran it locally by cloning the otel repo, replacing the packages with the local new versions, and ran the `make` command to no errors

songy23 reviewed (approved):
Approval for OTel-owned files

DylanLovesCoffee reviewed (approved):
LGTM for Serverless

zarirhamza commented:
/merge

zarirhamza committed:
add serverless dependencies

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
fix serverless dep typo

dustmop commented on cmd/serverless/dependencies_linux_amd64.txt:
```diff
@@ -919,6 +920,7 @@ html/template
 internal/abi
 internal/bisect
 internal/bytealg
+internal/chacha8rand
```
Remove this line to fix the `golang_deps_test` CI check

dustmop commented on cmd/serverless/dependencies_linux_arm64.txt:
```diff
@@ -918,6 +919,7 @@ html/template
 internal/abi
 internal/bisect
 internal/bytealg
+internal/chacha8rand
```
Remove this line to fix the `golang_deps_test` CI check

dustmop reviewed (approved):
LGTM for agent-shared-components

zarirhamza committed:
fix deps test

zarirhamza committed:
Merge branch 'main' into zarir/semantic-core-peer-tag-aggregation

zarirhamza committed:
remove newline at end of dep file

zarirhamza commented:
/merge
